bpf: Add unit tests for SNAT port allocation

gentoo-root opened the pull request:
The new test evaluates the effectiveness of the port allocation
algorithm in SNAT by simulating multiple clients connecting to multiple
servers via NAT. The clients use source port numbers sampled from a
Linux machine using the portgen.py script. Several assertions are then
made, based on metrics collected during the test.

Dummy NODEPORT_PORT_MAX_NAT is changed to reflect the actual production
value, so that the test could show realistic results.

bpf_loop helper is added to run hundreds of thousands of iterations.

```release-note
Unit tests for SNAT port allocation algorithm.
```

gentoo-root committed:
bpf: Add unit tests for SNAT port allocation

The new test evaluates the effectiveness of the port allocation
algorithm in SNAT by simulating multiple clients connecting to multiple
servers via NAT. The clients use source port numbers sampled from a
Linux machine using the portgen.py script. Several assertions are then
made, based on metrics collected during the test.

Dummy NODEPORT_PORT_MAX_NAT is changed to reflect the actual production
value, so that the test could show realistic results.

bpf_loop helper is added to run hundreds of thousands of iterations.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

borkmann reviewed (approved):
LGTM - I presume the first commit gets taken out and is merged via https://github.com/cilium/cilium/pull/36730 .

gentoo-root committed:
cocci: Handle array-style assignments to non-const pointers

When a pointer is passed to the function arguments, and it's written to
using the array syntax `p[i] = ...`, don't enforce const on that
pointer, because the code won't compile.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
cocci: Handle passing non-const pointers to designated initializers

When a pointer is passed to the function arguments, and it's further
passed to a struct field via a designated initializer, don't enforce
const on that pointer, because the target field is likely non-const.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

bimmlerd reviewed (commented):
could you add a CODEOWNERS entry for the `contrib/scripts/portgen.py` script and assign it to `@cilium/sig-datapath`? Otherwise it's owned by `@cilium/contributing` and we have little context on it :)
